File: nodejs/02-files.js
/*
 * This tutorial shows basic file operations in Zbox.
 */

const Zbox = require('@zbox/nodejs');

(async function run() {
  // create a Zbox instance
  const zbox = new Zbox();

  // initialise Zbox environment and turn on debug logs
  await zbox.initEnv({ debug: true });

  // open the repo
  var repo = await zbox.openRepo({
    uri: '[your_repo_uri]',
    pwd: 'secret password',
    opts: { create: true }
  });

  var file;

  // --------------------------------------------
  // create a file and write/read string
  // --------------------------------------------
  file = await repo.createFile('/file.txt');

  // write string to file, this will create a permanent version
  await file.writeOnce('Zbox is awesome!');

  // seek to the beginning of file
  await file.seek({ from: Zbox.SeekFrom.Start, offset: 0 });

  // read all string content
  const str = await file.readAllString()
  console.log(str);

  // close file
  await file.close();

  // --------------------------------------------
  // create a file and write/read binary
  // --------------------------------------------
  file = await repo.createFile('/file.bin');

  // write binary to file, this will create a permanent version
  const buf = new Uint8Array([1, 2, 3]);
  await file.writeOnce(buf.slice());

  // seek to the beginning of file
  await file.seek({ from: Zbox.SeekFrom.Start, offset: 0 });

  // read all binary content
  const result = await file.readAll();
  console.log(result);

  // close file
  await file.close();

  // --------------------------------------------
  // remove files
  // --------------------------------------------
  await repo.removeFile('/file.txt');
  await repo.removeFile('/file.bin');

  // close repo and exit Zbox
  await repo.close();
  await zbox.exit();
})();

File: rust/src/bin/02-files.rs
/*
 * This tutorial shows basic file operations in Zbox.
 */

extern crate zbox;

use std::io::{Read, Seek, SeekFrom};
use zbox::{init_env, RepoOpener};

fn main() {
    // initialise zbox environment, called first
    init_env();

    // create and open a repository
    let mut repo = RepoOpener::new()
        .create(true)
        .open("[your_repo_uri]", "secret password")
        .unwrap();

    // --------------------------------------------
    // create a file and write/read string
    // --------------------------------------------
    {
        let mut file = repo.create_file("/file.txt").unwrap();

        // write string to file, this will create a permanent version
        file.write_once(b"Zbox is awesome!").unwrap();

        // seek to the beginning of file
        file.seek(SeekFrom::Start(0)).unwrap();

        // read all string content
        let mut content = String::new();
        file.read_to_string(&mut content).unwrap();
        println!("{}", content);
    }

    // --------------------------------------------
    // create a file and write/read binary
    // --------------------------------------------
    {
        let mut file = repo.create_file("/file.bin").unwrap();

        // write binary to file, this will create a permanent version
        file.write_once(&[1u8, 2u8, 3u8]).unwrap();

        // seek to the beginning of file
        file.seek(SeekFrom::Start(0)).unwrap();

        // read all binary content
        let mut content = Vec::new();
        file.read_to_end(&mut content).unwrap();
        println!("{:?}", content);
    }

    // --------------------------------------------
    // remove files
    // --------------------------------------------
    repo.remove_file("/file.txt").unwrap();
    repo.remove_file("/file.bin").unwrap();
}
